torch/_decomp/decompositions.py

```diff
@@ -1963,6 +1963,12 @@ def uniform(
     )
 
 
+@register_decomposition(aten.exponential_)
+def exponential_(self, rate=1, generator=None):
+    assert generator is None
+    return self.copy_(-1 / rate * torch.log1p(-torch.rand_like(self)))
```

Comment:
```suggestion
    return self.copy_(-1 / rate * torch.log(torch.rand_like(self)))
```
If x ~ U(0,1), 1-x ~ U(0,1).

Reply:
This can go to refs?

Reply:
Same for the other distributions, yep

Reply:
@lezcano this won't work with triton (and generally with any fast log approximation) if compute is properly done in fp32, this is explained in the eager code). Exponential distribution should not generate 0's because pdf at 0 is 0. Yet with half dtype fast log approximation would be truncated to 0:
```
In [28]: max_rand = torch.rand(10000000000, device="cuda").amax()

In [29]: def fn(x):
    ...:     return x.log().half()
    ...: 

In [30]: opt_fn = torch.compile(fn)
/scratch/ngimel/work/pytorch/torch/_dynamo/eval_frame.py:372: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled.Consider setting `torch.set_float32_matmul_precision('high')`
  warnings.warn(

In [31]: fn(max_rand)
Out[31]: tensor(-5.9605e-08, device='cuda:0', dtype=torch.float16) #fine, eager log doesn't use fast approximation)
In [33]: opt_fn(max_rand)
Out[33]: tensor(-0., device='cuda:0', dtype=torch.float16) #fast log approx
```

Reply:
Right. I guess we'd have similar issues even if we cast it to float?

Reply:
That's precisely because of cast to float, if we didn't, max_rand would be further away from 1. But we want to cast to float, because that's more precise and that's what eager does.

Reply:
Right, I see. Great catch! Let's go with `log1p` then.

Reply:
> Exponential distribution should not generate 0's because pdf at 0 is 0.

I see numpy and tensorflow uses this definition. But does pytorch as well? 

`f(x) = lambda * e^(-lambda * x), x >= 0 --vs-- x > 0`  If it's the former, `x >= 0` then exponential variate does not exclude 0, and `f(x=0) = lambda` while `F(x=0) = 0`. I had a quick empirical check and I think it's define for `x >= 0` ?   


```python
>>> torch.empty(100).exponential_(1.17549e+45)
tensor([1.4013e-45, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.4013e-45,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.4013e-45, 0.0000e+00,
        0.0000e+00, 1.4013e-45, 1.4013e-45, 2.8026e-45, 0.0000e+00, 0.0000e+00,
        1.4013e-45, 2.8026e-45, 2.8026e-45, 0.0000e+00, 0.0000e+00, 1.4013e-45,
        1.4013e-45, 1.4013e-45, 1.4013e-45, 1.4013e-45, 0.0000e+00, 1.4013e-45,
        5.6052e-45, 0.0000e+00, 2.8026e-45, 0.0000e+00, 5.6052e-45, 2.8026e-45,
        0.0000e+00, 1.4013e-45, 1.4013e-45, 1.4013e-45, 0.0000e+00, 0.0000e+00,
        2.8026e-45, 1.4013e-45, 1.4013e-45, 1.4013e-45, 1.4013e-45, 4.2039e-45,
        2.8026e-45, 1.4013e-45, 0.0000e+00, 1.4013e-45, 0.0000e+00, 5.6052e-45,
        1.4013e-45, 2.8026e-45, 0.0000e+00, 1.4013e-45, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 1.4013e-45, 5.6052e-45, 5.6052e-45, 2.8026e-45,
        1.4013e-45, 1.4013e-45, 1.4013e-45, 1.4013e-45, 5.6052e-45, 1.4013e-45,
        0.0000e+00, 0.0000e+00, 2.8026e-45, 1.4013e-45, 1.4013e-45, 1.4013e-45,
        4.2039e-45, 4.2039e-45, 1.4013e-45, 1.4013e-45, 2.8026e-45, 0.0000e+00,
        4.2039e-45, 1.4013e-45, 1.4013e-45, 0.0000e+00, 2.8026e-45, 1.4013e-45,
        1.4013e-45, 4.2039e-45, 0.0000e+00, 1.4013e-45, 4.2039e-45, 0.0000e+00,
        0.0000e+00, 2.8026e-45, 4.2039e-45, 0.0000e+00])
>>> 0 in torch.empty(100).exponential_(1.17549e+45) # try sampling more times
True
>>> 0 in torch.empty(100000000).exponential_() # even with the default lambda=1, with big enough sampling
True
```

Reply:
Or this is device specific definition of exponential ?

Reply:
The current formula you have implemented should do. This comment https://github.com/pytorch/pytorch/pull/91673#pullrequestreview-1246726928 and https://github.com/pytorch/pytorch/pull/91673#discussion_r1068780565 are still relevant though.

Reply:
Apparently cpu incorrectly implements exponential in eager, but cuda `exponential_` indeed doesn't contain zero, at least until large lambda would cause underflow:
```
In [5]: torch.empty(100000000, device="cuda").exponential_().min()
Out[5]: tensor(5.9605e-08, device='cuda:0')
```

Reply:
This is because CPU implementation uses MKL (https://github.com/pytorch/pytorch/blob/b8057aa16d376eefe081d852335541e2a7609c40/aten/src/ATen/native/cpu/DistributionKernels.cpp#L149). Let me check with MKL folks whether the VSL [`vRngExponential`](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/statistical-functions/random-number-generators/distribution-generators/continuous-distributions/vrngexponential.html) excludes zero or not. If it doesn't, we should fix the MKL-based implementation from PyTorch side. cc @CaoE

Reply:
The link above says `f(x)` is defined for `x >= a`, where `a` is a displacement parameter. 

https://github.com/pytorch/pytorch/blob/b8057aa16d376eefe081d852335541e2a7609c40/aten/src/ATen/native/cpu/DistributionKernels.cpp#L149-L150

Here `a` is 0, so probably that's why cpu generated 0.